TombEngine/Objects/TR3/Trap/ElectricCleaner.cpp

```diff
@@ -0,0 +1,380 @@
+					item.ItemFlags[1] &= ~(1 << 0);	// Turn off 1st bit for flagDoDetection.
+				else
+					item.ItemFlags[1] |= (1 << 0);	// Turn on 1st bit for flagDoDetection.
+			}
+		}
+
+		AnimateItem(&item);
+
+		int probedRoomNumber = GetCollision(&item).RoomNumber;
+		if (item.RoomNumber != probedRoomNumber)
+			ItemNewRoom(itemNumber, probedRoomNumber);
+
+		auto radius = Vector2(object.radius, object.radius);
+		AlignEntityToSurface(&item, radius);
+
+		SpawnElectricCleanerSparks(item);
+		ElectricCleanerToItemCollision(item);
+	}
+
+	bool IsNextSectorValid(ItemInfo& item, const Vector3& Dir)
+	{
+		GameVector detectionPoint = item.Pose.Position + Dir * BLOCK(1);
+		detectionPoint.RoomNumber = item.RoomNumber;
+
+		auto col = GetCollision(detectionPoint);
+
+		//Is a wall
+		if (col.Block->IsWall(detectionPoint.x, detectionPoint.z))
+			return false;
+
+		//Is it a sliding slope?
+		if (col.Position.FloorSlope)
+			return false;
+
+		if (abs(col.FloorTilt.x) == 0 && abs(col.FloorTilt.y) == 0) //Is a flat tile
+		{
+			//Is a 1 click step (higher or lower).
+			int distanceToFloor = abs(col.Position.Floor - item.Pose.Position.y);
+			if (distanceToFloor >= CLICK(1))
+				return false;
+		}
+		else //Is a slope tile
+		{
+			//Is a 2 click step (higher or lower).
+			int distanceToFloor = abs(col.Position.Floor - item.Pose.Position.y);
+			if (distanceToFloor > CLICK(2))
+				return false;
+
+			short slopeAngle = ANGLE(0.0f);
+
+			if (col.FloorTilt.x > 0)
+				slopeAngle = -ANGLE(90.0f);
+			else if (col.FloorTilt.x < 0)
+				slopeAngle = ANGLE(90.0f);
+
+			if (col.FloorTilt.y > 0 && col.FloorTilt.y > abs(col.FloorTilt.x))
+				slopeAngle = ANGLE(180.0f);
+			else if (col.FloorTilt.y < 0 && -col.FloorTilt.y > abs(col.FloorTilt.x))
+				slopeAngle = ANGLE(0.0f);
+
+			auto angleDir = FROM_RAD(atan2(Dir.x, Dir.z));
+			auto alignment = slopeAngle - angleDir;
+
+			//Is slope not aligned with the direction?
+			if ((alignment != 32768) && (alignment != 0) && (alignment != -32768))
+				return false;
+		}
+
+		//Is diagonal floor?
+		if (col.Position.DiagonalStep)
+			return false;
+
+		//Is ceiling (square or diagonal) high enough?
+		int distanceToCeiling = abs(col.Position.Ceiling - col.Position.Floor);	
+		int cleanerHeight = 1024; //TODO change it for the collision bounding box height.
+		if (distanceToCeiling < cleanerHeight)
+			return false;
+
+		//Is a non walkable tile? (So there is not any box)
+		if (col.Block->Box == NO_BOX)
+			return false;
+
+		//Is a blocked grey box (So it's an Isolated box)
+		if (g_Level.Boxes[col.Block->Box].flags & BLOCKABLE)
+			return false;
+		
+		//Is a stopper tile? (There is still a shatter object).
+		if (col.Block->Stopper)
+			return false;
+
+		//If nothing of that happened, then it must be a valid sector.
+		return true;
+	}
+
+	Vector3 ElectricCleanerSearchDirections(ItemInfo& item, const Vector3& Dir1, const Vector3& Dir2, const Vector3& Dir3)
```

Comment:
could it be possible to change variable names from Dir1, Dir2 ... to something meaningful? Like DirX, DirY etc

Reply:
No, DirX DirY would be more confusing:

The function is taking 3 directions, and they are called Dir1, Dir2 Dir3 because are the order.

1st it checks Dir1, if it's ok, returns it, if not, checks Dir2... etc...

And those Dirs, may be pointing north, south, east or west... the purpose of this SearchDirection is to can reuse the same functions but for different directions. So, any other name can not be linked to a cardinal direction, becaues that may change.

(You can see how they are called around line 124, each case order different directions, so I let Dir1,2,3 because that's something all have in common).